### DISCUSSION TITLE
YOURLS couldn't get MariaDB version

### DISCUSSION THREAD

**Discussion by Fustie** [YOURLS/YOURLS#3166] [Q&A] [answered]
So I've installed YOURLS by [this](https://linuxstans.com/how-to-install-yourls/) tutorial, but for Manjaro linux (someway similar to Arch) and all works properly, but I always receive "MySQL version is too old. Ask your server admin for an upgrade." error. And when I turned debug mode on, I see that YOURLS doesn't see my MariaDB version!
`Connected to database yourlsdb on 127.0.0.1 
SQL perform: SELECT option_name, option_value FROM yourls_options WHERE 1=1 (0.00033 s)
MySQL version: `
PHP Version: 8.0.13
MariaDB Version: 5.5.5-10.6.5-MariaDB

**Comment by LeoColomb** [comment, marked as answer]
@Fustie Indeed, this `5.5.5` is not expected. Could you check with you host the reason why it is present in the version?

**Reply by dgw** [reply]
I don't think that's the source of the problem. YOURLS should see that as just fine; it's the first version-number-like thing in the given MariaDB version string, and it's higher than the expected 5.0. The presence of a fake MySQL version 5.5.5 before the actual MariaDB version number seems to be [expected](https://stackoverflow.com/a/57497111/5991) if the client library is not MariaDB-aware (original source is a MariaDB project member, commenting on a Joomla issue).

The fact that YOURLS is outputting an empty MySQL version seems to point somewhere else. `yourls_get_database_version()` can be [overridden by a plugin filter](https://github.com/YOURLS/YOURLS/blob/42fd3b85c2d49120f3328c7f77ce4d5c100fcc61/includes/functions-install.php#L28-L32), for example. Or perhaps this particular server is configured in some way that the version fetched via PDO's method really is empty due to a bug somewhere else.

My point is, after poking around the relevant code twice today, I don't see how it can be an issue with the version number format. I'm confident that `yourls_sanitize_version()` would deal with this case if the input was as expected.

**Reply by Fustie** [reply]
It is MySQL version, I got this string from php.

**Reply by Fustie** [reply]
@dgw I installed fresh YOURLS, no plugins.

**Reply by dgw** [reply]
@Fustie You got the version string from PHP, where? Seems like maybe some issue in how PDO fetches the `PDO::ATTR_SERVER_VERSION` value on this system vs. how it normally works. I went down a rabbit hole for half an hour but got very lost in PDO's internals and ultimately didn't find any hints. MariaDB's source wasn't much help either, because I don't know where anything is. 😅

Perhaps it would be helpful to know if you can find which driver PDO is using, mysqli or mysqlnd. At least that would help narrow the search.

**Reply by Fustie** [reply]
@dgw I use this code to get MySQL version from PHP.
```
<?php
$link = mysqli_connect( 'localhost', 'mysql_user', 'mysql_password' );
var_dump( $link->server_info );
}
?>
```

**Reply by dgw** [reply]
OK, that means mysqli works, at least in theory. Could be interesting to compare the output of this:

```php
<?php
$db = new PDO('mysql:host=localhost', 'mysql_user', 'mysql_password');
var_dump($db->getAttribute(PDO::ATTR_DRIVER_NAME));
var_dump($db->getAttribute(PDO::ATTR_SERVER_VERSION));
```

If the version is empty and the driver name is something other than mysqli, that'll be a clue.

**Reply by Fustie** [reply]
I received this when issue your script `PHP message: PHP Parse error:  syntax error, unexpected token ";", expecting ")" in /usr/share/nginx/html/YOURLS/admin/lol.php on line 3`, also I found this error `PHP Warning:  PHP Startup: Unable to load dynamic library 'pdo_dblib' (tried: /usr/lib/php/modules/pdo_dblib (/usr/lib/php/modules/pdo_dblib: cannot open shared object file: No such file or directory)`

**Reply by dgw** [reply]
I forgot to close the `var_dump`s, whoops.

pdo_dblib should be irrelevant, as it appears to be for MSSQL servers.

**Reply by ozh** [reply]
Anyway : `5.5.5-10.6.5-MariaDB` is sanitized by YOURLS as `5.5.5` which should anyway pass since [we're checking for 5.0](https://github.com/YOURLS/YOURLS/blob/master/includes/functions-install.php#L13-L19)+

Simple fix if you don't want to dig further : make a simple plugin to hook into `shunt_get_database_version` and return `10.0.0` -- see https://github.com/YOURLS/YOURLS/blob/42fd3b85c2d49120f3328c7f77ce4d5c100fcc61/includes/functions-install.php#L21-L35

**Comment by Fustie** [comment]
For some reason, all start work after reboot of server, thank you all!
